In [19]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")  # Ignore all warnings

In [20]:
df = pd.read_excel('../data/customer-call-list.xlsx')
df.head(3)

CustomerID First_Name Last_Name  Phone_Number                Address  \
0        1001      Frodo   Baggins  123-545-5421  123 Shire Lane, Shire   
1        1002       Abed     Nadir  123/643/9775    93 West Main Street   
2        1003     Walter    /White    7066950392     298 Drugs Driveway   

  Paying Customer Do_Not_Contact  Not_Useful_Column  
0             Yes             No               True  
1              No            Yes              False  
2               N            NaN               True

In [21]:
# df.dtypes
df.shape

(21, 8)

## The data is a **customer-call-list** containing information about customers

### The intended use of the data is for decision-making purposes

### Data Format Issues:

- Phone_Number: The format is inconsistent with a mix of hyphens, slashes, and pipes as delimiters.
- Address: Some addresses lack zip codes or have state information embedded within the address field.
- Paying Customer: Values are inconsistent, using both "Yes/No" and "Y/N" variations.
- Do_Not_Contact: Similar to "Paying Customer," uses both "Yes/No" and "Y/N" variations. Also, some entries are blank, which may need to be treated as a separate category or standardized.
- Not_Useful_Column: This column seems irrelevant and can be removed.

### Data Quality Issues:
- Duplicate Entry: CustomerID 1020 appears twice with identical information.
- Missing Values: Several fields have missing values, notably in Last_Name for customer 1003 and Phone_Number for customer 1007.
- Inconsistent Naming: First and last names sometimes include extra spaces (e.g., Jeff Winger, Michael Scott).

### Recommendations:
#### 1. Data Cleaning:
- Standardize phone number format to improve data quality.
- Separate state and zip code in the address field.
- Unify "Paying Customer" and "Do_Not_Contact" values to either "Yes/No" or "Y/N".
- Address missing values through further investigation or appropriate imputation methods.
- Remove extra spaces from names and ensure consistent capitalization.
- Remove duplicate entry for customer 1020.
- Delete the "Not_Useful_Column".

#### 2. Data Validation:
- Implement data validation rules to prevent future inconsistencies.
- For instance, enforce format restrictions on phone numbers and ensure mandatory fields are filled.

#### Additional Notes:
- By addressing these issues, you can improve the quality and reliability of the customer data for better decision-making and analysis.

In [22]:
# # Steps
# remove duplicates
# drop columns
# standardize the data
# Null values or blank values
# remove any columns

In [23]:
# # BEST PRACTICE 
# create a copy of the dataframe
# df_copy = df.copy()
# df_copy.head(3)

In [24]:
# Remove unnecessary column
df = df.drop('Not_Useful_Column', axis=1) # axis = 1 means column-wise drop, axis = 0 means row-wise drop
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N  
20            910 Tatooine Road, Tatooine             Yes              N

In [25]:
# drop duplicates
df = df.drop_duplicates()
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [26]:
# df['Last_Name'].str.strip() # strip() removes leading and trailing WHITE spaces
df['Last_Name'] = df['Last_Name'].str.strip('./_') # strip() removes specified characters ('./123')
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [27]:
def clean_phone_number(phone_number):
    """Cleans and standardizes phone numbers.

    Args:
        phone_number: The phone number to clean.

    Returns:
        The cleaned phone number in the format "XXX-XXX-XXXX", or None if the
        input is missing or invalid.
    """
    if pd.isna(phone_number) or str(phone_number).lower() in ("n/a", "na"):
        return None

    # Remove non-numeric characters and plus signs
    digits = ''.join(c for c in str(phone_number) if c.isdigit() or c == '+')

    # Ensure we have at least 10 digits (assuming North American Numbering Plan)
    if len(digits) < 10:
        return None

    # Handle phone numbers with leading plus sign
    if digits.startswith('+'):
        return f"{digits[1:4]}-{digits[4:7]}-{digits[7:]}"
    else:
        return f"{digits[:3]}-{digits[3:6]}-{digits[6:]}"

# Clean phone numbers
df['Phone_Number'] = df['Phone_Number'].apply(clean_phone_number)
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White  706-695-0392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger          None   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN          None   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee          None   
11        1012      Harry      Potter  706-695-0392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent  706-695-0392   
18        1019      Creed      Braton          None   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [28]:
# Split the 'Address' column into multiple columns
new_columns = ['Street_Address', 'State', 'Zip_Code']  # Customize column names as needed
df[new_columns] = df['Address'].str.split(',', expand=True)
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White  706-695-0392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger          None   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN          None   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee          None   
11        1012      Harry      Potter  706-695-0392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent  706-695-0392   
18        1019      Creed      Braton          None   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire     None  
1    93 West Main Street           None     None  
2     298 Drugs Driveway           None     None  
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road           None     None  
5       768 City Parkway           None     None  
6      1209 South Street           None     None  
7          98 Clue Drive           None     None  
8       123 Middle Earth           None     None  
9       25th Main Street       New York     None  
10        612 Shire Lane          Shire     None  
11  2394 Hogwarts Avenue           None     None  
12      2039 Main Street           None     None  
13      343 City Parkway           None     None  
14         214 HR Avenue           None     None  
15  2395 Hogwarts Avenue           None     None  
16      121 Paper Avenue   Pennsylvania     None  
17       3498 Super Lane           None     None  
18                   N/a           None     None  
19     910 Tatooine Road       Tatooine     None

In [29]:
# drop the original 'Address' column
df = df.drop('Address', axis=1)
df

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421             Yes   
1         1002       Abed       Nadir  123-643-9775              No   
2         1003     Walter       White  706-695-0392               N   
3         1004     Dwight     Schrute  123-543-2345             Yes   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467             Yes   
6         1007       Jeff      Winger          None              No   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN          None             Yes   
9         1010      Peter      Parker  123-545-5421             Yes   
10        1011    Samwise      Gamgee          None             Yes   
11        1012      Harry      Potter  706-695-0392               Y   
12        1013        Don      Draper  123-543-2345             Yes   
13        1014     Leslie       Knope  876-678-3469             Yes   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421              No   
16        1017   Michael        Scott  123-643-9775             Yes   
17        1018      Clark        Kent  706-695-0392               Y   
18        1019      Creed      Braton          None             N/a   
19        1020     Anakin   Skywalker  876-678-3469             Yes   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

In [30]:
df['Paying Customer'] = df['Paying Customer'].str.replace('Yes', 'Y').str.replace('No', 'N')
df

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White  706-695-0392               N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger          None               N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN          None               Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee          None               Y   
11        1012      Harry      Potter  706-695-0392               Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent  706-695-0392               Y   
18        1019      Creed      Braton          None             N/a   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

In [31]:
df['Do_Not_Contact'] = df['Do_Not_Contact'].str.replace('Yes', 'Y').str.replace('No', 'N')
df

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White  706-695-0392               N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger          None               N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN          None               Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee          None               Y   
11        1012      Harry      Potter  706-695-0392               Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent  706-695-0392               Y   
18        1019      Creed      Braton          None             N/a   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire     None  
1               Y   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4               N      123 Dragons Road           None     None  
5               Y      768 City Parkway           None     None  
6               N     1209 South Street           None     None  
7               N         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9               N      25th Main Street       New York     None  
10              N        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13              N      343 City Parkway           None     None  
14              N         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16              N      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18              Y                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

In [32]:
df = df.fillna('')
df

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White  706-695-0392               N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf                                         Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter  706-695-0392               Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent  706-695-0392               Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
1               Y   93 West Main Street                          
2                    298 Drugs Driveway                          
3               Y      980 Paper Avenue   Pennsylvania    18503  
4               N      123 Dragons Road                          
5               Y      768 City Parkway                          
6               N     1209 South Street                          
7               N         98 Clue Drive                          
8                      123 Middle Earth                          
9               N      25th Main Street       New York           
10              N        612 Shire Lane          Shire           
11                 2394 Hogwarts Avenue                          
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
17                      3498 Super Lane                          
18              Y                   N/a                          
19              N     910 Tatooine Road       Tatooine

In [33]:
for x in df.index:
    if df.loc[x, 'Do_Not_Contact'] == 'Y':
        df.drop(x, inplace=True)
df
    

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
2         1003     Walter       White  706-695-0392               N   
4         1005        Jon        Snow  876-678-3469               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf                                         Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter  706-695-0392               Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent  706-695-0392               Y   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
2                    298 Drugs Driveway                          
4               N      123 Dragons Road                          
6               N     1209 South Street                          
7               N         98 Clue Drive                          
8                      123 Middle Earth                          
9               N      25th Main Street       New York           
10              N        612 Shire Lane          Shire           
11                 2394 Hogwarts Avenue                          
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
17                      3498 Super Lane                          
19              N     910 Tatooine Road       Tatooine

In [34]:
for x in df.index:
    if df.loc[x, 'Phone_Number'] == '':
        df.drop(x, inplace=True)
df

# another way to drop NULL values
# df.dropna(subset=['Phone_Number'], inplace=True)

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
2         1003     Walter       White  706-695-0392               N   
4         1005        Jon        Snow  876-678-3469               Y   
7         1008   Sherlock      Holmes  876-678-3469               N   
9         1010      Peter      Parker  123-545-5421               Y   
11        1012      Harry      Potter  706-695-0392               Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent  706-695-0392               Y   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
2                    298 Drugs Driveway                          
4               N      123 Dragons Road                          
7               N         98 Clue Drive                          
9               N      25th Main Street       New York           
11                 2394 Hogwarts Avenue                          
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
17                      3498 Super Lane                          
19              N     910 Tatooine Road       Tatooine

In [36]:
# reset index
df.reset_index(drop=True)

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1003     Walter       White  706-695-0392               N   
2         1005        Jon        Snow  876-678-3469               Y   
3         1008   Sherlock      Holmes  876-678-3469               N   
4         1010      Peter      Parker  123-545-5421               Y   
5         1012      Harry      Potter  706-695-0392               Y   
6         1013        Don      Draper  123-543-2345               Y   
7         1014     Leslie       Knope  876-678-3469               Y   
8         1015       Toby  Flenderson  304-762-2467               N   
9         1016        Ron     Weasley  123-545-5421               N   
10        1017   Michael        Scott  123-643-9775               Y   
11        1018      Clark        Kent  706-695-0392               Y   
12        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
1                    298 Drugs Driveway                          
2               N      123 Dragons Road                          
3               N         98 Clue Drive                          
4               N      25th Main Street       New York           
5                  2394 Hogwarts Avenue                          
6               N      2039 Main Street                          
7               N      343 City Parkway                          
8               N         214 HR Avenue                          
9               N  2395 Hogwarts Avenue                          
10              N      121 Paper Avenue   Pennsylvania           
11                      3498 Super Lane                          
12              N     910 Tatooine Road       Tatooine